### PR TITLE
Add TESTING.md and BENCHMARKING.md

### PR DESCRIPTION
**Pull request by harrism**
## Description
Contributes to #598. Adds `cpp/doc/TESTING.md` and `cpp/doc/BENCHMARKING.md`, the remaining parts of the libcuspatial C++ developer documentation. These documents cover unit testing and unit benchmarking best practices, respectively.

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cuspatial/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

File: cpp/doc/BENCHMARKING.md [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+Both NVBench and Google Benchmark provide many options for specifying ranges of parameters to
+benchmark, as well as to control the time unit reported, among other options. Refer to existing
+benchmarks in `cpp/benchmarks` to understand the options.
+
+## Directory and File Naming
+
+The naming of unit benchmark directories and source files should be consistent with the feature
+being benchmarked. For example, the benchmarks for APIs in `point_in_polygon.hpp` should live in
+`cpp/benchmarks/point_in_polygon.cu`. Each feature (or set of related features) should have its own
+benchmark source file named `<feature>.cu/cpp`.
```
Comment (thomcom) [inline cpp/doc/BENCHMARKING.md:20]:
```suggestion
benchmark source file named `<feature>{.cu/cpp}`. 
```

File: cpp/doc/BENCHMARKING.md [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+In general, we should benchmark all features over a range of data sizes and types, so that we can
+catch regressions across libcudf changes. However, running many benchmarks is expensive, so ideally
+we should sample the parameter space in such a way to get good coverage without having to test
+exhaustively.
+
+A rule of thumb is that we should benchmark with enough data to reach the point where the algorithm
+reaches its saturation bottleneck, whether that bottleneck is bandwidth or computation. Using data
+sets larger than this point is generally not helpful, except in specific cases where doing so
+exercises different code and can therefore uncover regressions that smaller benchmarks will not
+(this should be rare).
```
Comment (thomcom) [inline cpp/doc/BENCHMARKING.md:55]: Do think we should suggest here that benchmarks should only be written for public APIs?

Comment (thomcom) [inline cpp/doc/BENCHMARKING.md:55]: Do think we should suggest here that benchmarks should only be written for public APIs?

File: cpp/doc/TESTING.md [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+
+ * Most algorithms should have one or more tests exercising inputs with a large enough number of
+   rows to require launching multiple thread blocks, especially when values are ultimately
+   communicated between blocks (e.g. reductions). This is especially important for custom kernels
+   but also applies to Thrust and CUB algorithm calls with lambdas / functors.
+
+ * Test exceptional cases. For example, anything that causes the function to `throw`.
+
+ * Test boundary cases. For example points that fall exactly on lines or boundaries.
+
```
Comment (thomcom) [inline cpp/doc/TESTING.md:23]: Similar to the above, do we want to specify tests written only for public/non-`detail` APIs, but that those tests should be written such that as close to 100% test coverage of internal functions occurs?

Comment (thomcom) [reply cpp/doc/TESTING.md:23]: This reminds me that `detail` APIs typically include the `mr` and `stream` args which I don't think we are currently testing. Do we want to mention this?

File: cpp/doc/TESTING.md [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+using TestTypes = ::test::types<float,double>; // Notice custom cudf type list type
+TYPED_TEST_SUITE(TypedTestFixture, TestTypes);
+TYPED_TEST(TypedTestFixture, FirstTest){
+    // Access the current type using `TypeParam`
+    using T = TypeParam;
+}
+```
+
+In this example, all tests using the `TypedTestFixture` fixture will run once for each type in the
+list defined in `TestTypes` (`int, float, double`).
```
Comment (thomcom) [inline cpp/doc/TESTING.md:82]:
```suggestion
list defined in `TestTypes` (`float, double`).
```

**Review by thomcom — CHANGES_REQUESTED** [review]

File: cpp/doc/TESTING.md (hunk 4 above)
Comment (thomcom) [reply cpp/doc/TESTING.md:82]: This is the only mistake that I saw, everything else looks good to approve.

File: cpp/doc/TESTING.md (hunk 3 above)
Comment (harrism) [reply cpp/doc/TESTING.md:23]: Public APIs have MR args if they allocate output memory, so that is tested. All header-only APIs have non-optional stream args. We are moving most testing to the header-only API.

File: cpp/doc/BENCHMARKING.md [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+The naming of unit benchmark directories and source files should be consistent with the feature
+being benchmarked. For example, the benchmarks for APIs in `point_in_polygon.hpp` should live in
+`cpp/benchmarks/point_in_polygon.cu`. Each feature (or set of related features) should have its own
+benchmark source file named `<feature>{.cu,cpp}`. 
+
+In the interest of improving compile time, whenever possible, test source files should be `.cpp`
+files because `nvcc` is slower than `gcc` in compiling host code. Note that `thrust::device_vector`
+includes device code, and so must only be used in `.cu` files. `rmm::device_uvector`,
+`rmm::device_buffer` and the various `column_wrapper` types described in [Testing](TESTING.md)
+can be used in `.cpp` files, and are therefore preferred in test code over `thrust::device_vector`.
```
Comment (isVoid) [inline cpp/doc/BENCHMARKING.md:26]: Does using NVBench precludes naming the file as `cpp`? NVBench header is `cuh`.

File: cpp/doc/TESTING.md [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+ * Test exceptional cases. For example, anything that causes the function to `throw`.
+
+ * Test boundary cases. For example points that fall exactly on lines or boundaries.
+
+## Header-only and Column-based API tests
+
+libcuspatial currently has two C++ APIs: the column-based API uses libcudf data structures as 
+input and output. These tests live in `cpp/tests/` and can use libcudf features for constructing
+columns and tables. The header-only API does not depend on libcudf at all and so tests of these
+APIs should not include any libcudf headers. These tests currently live in `cpp/tests/experimental`.
```
Comment (isVoid) [inline cpp/doc/TESTING.md:29]: What's unstated here is that we generally leave all "core" test cases (test cases that's specific to computation) to the experimental API, and "peripheral" test cases (test cases that's specific to column API on type handling, input validation etc.) to the column API.

The compile time requirements on the experimental APIs are generally not tested. Because they don't pass compile.

File: cpp/doc/BENCHMARKING.md (hunk 5 above)
Comment (harrism) [reply cpp/doc/BENCHMARKING.md:26]: Never thought about that. Asking.

File: cpp/doc/TESTING.md (hunk 3 above)
Comment (harrism) [reply cpp/doc/TESTING.md:23]: Done.

File: cpp/doc/BENCHMARKING.md (hunk 2 above)
Comment (harrism) [reply cpp/doc/BENCHMARKING.md:55]: Done.

File: cpp/doc/TESTING.md (hunk 6 above)
Comment (harrism) [reply cpp/doc/TESTING.md:32]: Done.

**Comment by harrism** [comment]
rerun tests

File: cpp/doc/BENCHMARKING.md (hunk 5 above)
Comment (harrism) [reply cpp/doc/BENCHMARKING.md:26]: I think nvbench is .cu only...

Comment (harrism) [reply cpp/doc/BENCHMARKING.md]: Removed this paragraph.

**Comment by harrism** [comment]
@gpucibot merge
